torch/csrc/jit/custom_operator.h

```diff
@@ -0,0 +1,190 @@
+  auto arguments = createArgumentVectorFromTraits<FunctionTraits>(
+      typename MakeIndices<FunctionTraits::number_of_parameters>::indices{});
+  auto returns = createReturns(static_cast<ReturnType*>(nullptr));
+  return {name, arguments, returns};
+}
+
+/// Does two things for an operator implementation and a tuple of arguments:
+/// 1. Pops all necessary arguments off the stack into the tuple's elements,
+/// 2. Unpacks the tuple and calls the operator implementation.
+/// The result of the implementation call is returned.
+template <
+    typename ReturnType,
+    typename Implementation,
+    typename... Types,
+    size_t... Is>
+ReturnType callOperatorWithTuple(
+    Implementation&& implementation,
+    Stack& stack,
+    std::tuple<Types...>& tuple,
+    Indices<Is...>) {
+  pop(stack, std::get<Is>(tuple)...);
+  return std::forward<Implementation>(implementation)(std::get<Is>(tuple)...);
+}
+
+void checkArgumentVector(
+    const char* what,
+    const std::vector<Argument>& inferred,
+    const std::vector<Argument>& provided,
+    const FunctionSchema& inferredSchema,
+    const FunctionSchema& providedSchema) {
+  AT_CHECK(
+      inferred.size() == provided.size(),
+      "Inferred ", inferred.size(), " ", what,
+      "(s) for operator implementation, but the provided schema specified ",
+      provided.size(), " ", what, "(s). Inferred schema: ",
+      inferredSchema, " | Provided schema: ", providedSchema);
+  for (size_t i = 0; i < provided.size(); ++i) {
+    AT_CHECK(
+        provided[i].type == inferred[i].type,
+        "Inferred type for ", what, " #", i, " was ",
+        *inferred[i].type, ", but the provided schema specified type ",
+        *provided[i].type, " for the ", what,
+        " in that position. Inferred schema: ",
+        inferredSchema, " | Provided schema: ", providedSchema);
+  }
+}
+
+/// If `schemaOrName` contains a `(`, it is assumed it specifies a schema, else
+/// it is assumed it only specifies the name. In the case where it is a full
+/// schema (assumed), we nevertheless infer the schema and verify that the user
+/// made no mistakes. Either way, this function returns the final schema.
+template <typename Traits>
+FunctionSchema inferAndCheckSchema(const std::string& schemaOrName) {
+  // If there is no '(' in the schema, we assume this is only the name (e.g.
+  // "foo::bar").
+  const auto bracketIndex = schemaOrName.find('(');
+  if (bracketIndex == std::string::npos) {
+    // Infer the full schema and we're good.
+    return torch::jit::detail::createFunctionSchemaFromTraits<Traits>(
+        /*name=*/schemaOrName);
+  }
+
+  // If the user provided her own schema, we need to infer it nevertheless and
+  // check that it's correct. We return the user provided schema in the end
+  // because it has proper argument names.
+
+  auto providedSchema = parseSchema(schemaOrName);
+  // Pick out only the name part.
+  auto name = schemaOrName.substr(0, bracketIndex);
```

Comment:
nit: 'aten::foo               \t\n(' is completely legit here so you probably want to use `providedSchema.name` instead.